### PR TITLE
*: Show key content in DuplicateEntry error message

### PR DESCRIPTION
**Pull request by shenli**
Fix: https://github.com/pingcap/tidb/issues/822

### CODE REVIEW

File: kv/union_store.go [hunk 1]
```diff
@@ -141,45 +154,36 @@ func (us *unionStore) Get(k Key) ([]byte, error) {
 }
 
 // markLazyConditionPair marks a kv pair for later check.
-func (us *unionStore) markLazyConditionPair(k Key, v []byte) error {
-	if len(v) == 0 {
-		return errors.Trace(us.lazyConditionPairs.Delete(k))
+func (us *unionStore) markLazyConditionPair(k Key, v []byte, e error) error {
+	us.lazyConditionPairs[string(k)] = &conditionPair{
+		key:   k,
```
Comment (disksing) [inline kv/union_store.go]: The Key should be cloned before store in map, the caller may change the content after calling `Get()`.
Or maybe we can use the map key directly in later check, then no need to store it here.

File: kv/union_store.go [hunk 2]
```diff
@@ -141,45 +154,36 @@ func (us *unionStore) Get(k Key) ([]byte, error) {
 }
 
 // markLazyConditionPair marks a kv pair for later check.
-func (us *unionStore) markLazyConditionPair(k Key, v []byte) error {
-	if len(v) == 0 {
-		return errors.Trace(us.lazyConditionPairs.Delete(k))
+func (us *unionStore) markLazyConditionPair(k Key, v []byte, e error) error {
```
Comment (coocood) [inline kv/union_store.go]: Can we remove the return error?

File: kv/union_store.go [hunk 3]
```diff
@@ -121,7 +129,12 @@ func (us *unionStore) Get(k Key) ([]byte, error) {
 	v, err := us.MemBuffer.Get(k)
 	if IsErrNotFound(err) {
 		if _, ok := us.opts.Get(PresumeKeyNotExists); ok {
-			err = us.markLazyConditionPair(k, nil)
+			e, ok := us.opts.Get(PresumeKeyNotExistsError)
+			if ok {
+				err = us.markLazyConditionPair(k, nil, e.(error))
```
Comment (disksing) [inline kv/union_store.go]: Should we check the type conversion here in order to make sure it will not panic?

Comment (shenli) [inline kv/union_store.go]: We should only put error into the entry with PresumeKeyNotExistsError as key.
But may be we should check if e is nil.

File: kv/union_store.go [hunk 4]
```diff
@@ -141,45 +154,36 @@ func (us *unionStore) Get(k Key) ([]byte, error) {
...
-	it.Close()
-
-	if len(keys) == 0 {
+	if len(us.lazyConditionPairs) == 0 {
 		return nil
 	}
-	values, err := us.snapshot.BatchGet(keys)
-	if err != nil {
-		return errors.Trace(err)
+	var keys []Key
```
Comment (disksing) [inline kv/union_store.go]: This slice can be made with capacity of the map's length.

File: table/tables/tables.go [hunk 5]
```diff
@@ -416,12 +416,15 @@ func (t *Table) AddRecord(ctx context.Context, r []interface{}) (recordID int64,
...
+		e := kv.ErrKeyExists.Gen("Duplicate entry '%d' for key 'PRIMARY'", recordID)
+		txn.SetOption(kv.PresumeKeyNotExistsError, e)
 		_, err = txn.Get(recordKey)
 		if err == nil {
-			return recordID, kv.ErrKeyExists
+			return recordID, errors.Trace(e)
 		} else if !terror.ErrorEqual(err, kv.ErrNotExist) {
 			return 0, errors.Trace(err)
 		}
+		txn.DelOption(kv.PresumeKeyNotExistsError)
```
Comment (disksing) [inline table/tables/tables.go]: I think it is better to use a `defer` to delete option, or the options value will be left undeleted if error occurs.

**Comment by disksing** [comment]
LGTM

**Comment by shenli** [comment]
@coocood  PTAL

**Comment by coocood** [comment]
LGTM

**Pull request merged**